### PR TITLE
added debugger support

### PR DESCRIPTION
**Pull request by bharadwaj-aditya**
Added support for running Harbour Bridge using the debugger on vscode 

- [x] Tests pass
- [x] Appropriate changes to README are included in PR

### CODE REVIEW

**Comment by bharadwaj-aditya** [comment]
@googlebot I signed it!

**Comment by bharadwaj-aditya** [comment]
@agasheesh - please review

**Comment by bharadwaj-aditya** [comment]
@nevinheintze - please review

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
+{
+    // Use IntelliSense to learn about possible attributes.
+    // Hover to view descriptions of existing attributes.
+    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
+    "version": "0.2.0",
+    "configurations": [
+
```
Comment (agasheesh) [inline .vscode/launch.json:7]: stray empty line?

File: main.go [hunk 2]
```diff
@@ -58,6 +59,7 @@ func init() {
 	flag.BoolVar(&skipForeignKeys, "skip-foreign-keys", false, "skip-foreign-keys: if true, skip creating foreign keys after data migration is complete (ddl statements for foreign keys can still be found in the downloaded schema.ddl.txt file and the same can be applied separately)")
 	flag.StringVar(&sessionJSON, "session", "", "session: specifies the file we restore session state from (used in schema-only to provide schema and data mapping)")
 	flag.BoolVar(&webapi, "web", false, "web: run the web interface (experimental)")
+	flag.StringVar(&dumpFile, "dump-file", "", "location of dump file to process")
```
Comment (agasheesh) [inline main.go:62]: nit; suggest rewording the help text as "dump-file: location of dump file to process" to keep it consistent with other flags defined above

File: main.go [hunk 3]
```diff
@@ -98,7 +100,9 @@ func main() {
 		panic(fmt.Errorf("can't use both schema-only and skip-foreign-keys at once. Foreign Key creation can only be skipped when data migration takes place."))
 	}
 
-	ioHelper := &conversion.IOStreams{In: os.Stdin, Out: os.Stdout}
+	input := loadInput(dumpFile)
+
```
Comment (agasheesh) [inline main.go:104]: nit; stray empty space?

File: main.go [hunk 4]
```diff
@@ -144,3 +148,16 @@ func main() {
 		panic(err)
 	}
 }
+
+func loadInput(dumpFile string) *os.File {
+	if dumpFile != "" {
```
Comment (agasheesh) [inline main.go:153]: nit; a comment something like following might be useful - "we read the dumpFile specified by the user; else we expect to read dumpFile from standard input"

File: main.go [hunk 5]
```diff
@@ -98,7 +100,9 @@ func main() {
 		panic(fmt.Errorf("can't use both schema-only and skip-foreign-keys at once. Foreign Key creation can only be skipped when data migration takes place."))
 	}
 
-	ioHelper := &conversion.IOStreams{In: os.Stdin, Out: os.Stdout}
+	input := loadInput(dumpFile)
```
Comment (agasheesh) [inline main.go:103]: nit; should we add some defensive code to check that dumpFile is passed in only in the case of driver being mysqldump or pgdump?

File: main.go [hunk 6]
```diff
@@ -144,3 +148,16 @@ func main() {
 		panic(err)
 	}
 }
+
+func loadInput(dumpFile string) *os.File {
+	if dumpFile != "" {
+		fmt.Printf("\nloading dump file from path: %s\n", dumpFile)
```
Comment (agasheesh) [inline main.go:154]: nit; Suggest uppercasing first letter i.e., "\nLoading ..." since this will print on a new line

File: main.go [hunk 7]
```diff
@@ -44,6 +44,7 @@ var (
 	skipForeignKeys  bool
 	sessionJSON      string
 	webapi           bool
+	dumpFile         string
```
Comment (agasheesh) [inline main.go:47]: nit; should we call this dumpFilePath?

File: main.go [hunk 8]
```diff
@@ -144,3 +148,16 @@ func main() {
...
 	}
 }
+
+func loadInput(dumpFile string) *os.File {
+	if dumpFile != "" {
+		fmt.Printf("\nloading dump file from path: %s\n", dumpFile)
+		file, err := os.Open(dumpFile)
+		if err != nil {
+			fmt.Printf("\nerror reading file: %v err:%v", dumpFile, err)
+			panic(fmt.Errorf("cannot open file"))
```
Comment (agasheesh) [inline main.go:158]: consider passing err itself to panic since it will capture more useful details like permission error, file not existent etc; not sure if we still need fmt.Printf in this case?

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -26,3 +26,12 @@ information on using pull requests.
...
 This project follows
 [Google's Open Source Community Guidelines](https://opensource.google/conduct/).
+
+## IDE and Debugger
+
+We support VSCode out of the box for debugging the project.
+Follow these steps to use the debugger
+1. Setup the [Delve debugger](https://github.com/golang/vscode-go)
+2. Add the command line arguments in the [launch.json](https://github.com/cloudspannerecosystem/harbourbridge/blob/master/.vscode/launch.json)
+3. Run the project from main.go
```
Comment (agasheesh) [inline CONTRIBUTING.md:36]: Thanks for adding this, I tried it out and it looked super useful!

File: .vscode/launch.json [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+        {
+            "name": "Launch Package",
+            "type": "go",
+            "request": "launch",
+            "mode": "auto",
+            "program": "${fileDirname}",
+            "args": ["-driver=mysqldump", "-instance=e2e-spanner-testing", "-v=true", "-dump-file="]
+        }
+    ]
+}
```
Comment (agasheesh) [inline .vscode/launch.json:17]: nit; Does this need a new line to get rid of this red warning sort of?

File: .vscode/launch.json [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+    "version": "0.2.0",
+    "configurations": [
+
+        {
+            "name": "Launch Package",
+            "type": "go",
+            "request": "launch",
+            "mode": "auto",
+            "program": "${fileDirname}",
+            "args": ["-driver=mysqldump", "-instance=e2e-spanner-testing", "-v=true", "-dump-file="]
```
Comment (agasheesh) [inline .vscode/launch.json:14]: suggest reformatting this as to make it easier to edit

```
"args": [
    "-driver=mysqldump", 
    "-instance=e2e-spanner-testing", 
    "-v=true",
    "-dump-file="
]
```

File: .vscode/launch.json (hunk 1 above)
Comment (bharadwaj-aditya) [reply .vscode/launch.json:7]: removed

File: main.go (hunk 2 above)
Comment (bharadwaj-aditya) [reply main.go:62]: done

File: main.go (hunk 3 above)
Comment (bharadwaj-aditya) [reply main.go:104]: removed

File: main.go (hunk 4 above)
Comment (bharadwaj-aditya) [reply main.go:153]: added

File: main.go (hunk 7 above)
Comment (bharadwaj-aditya) [reply main.go:47]: makes sense. Changed.
I have left the flag as is , calling it dump-file as that seemed more intuitive as a command line parameter

File: main.go (hunk 5 above)
Comment (bharadwaj-aditya) [reply main.go:103]: Based on the current handling of the other flags i think this is fine. Adding conditional code here for a specific case seems a bit odd to me.

I think we should handle this as a part of the SubCommands change you are taking up, wherein we should take a call on whether we should ignore parameters that are not relevant to the current flow or to raise an error.

File: main.go (hunk 8 above)
Comment (bharadwaj-aditya) [reply main.go:158]: Just followed the template of the other panic blocks.
In any case, changed here.

File: .vscode/launch.json (hunk 11 above)
Comment (bharadwaj-aditya) [reply .vscode/launch.json]: changed

File: .vscode/launch.json (hunk 10 above)
Comment (bharadwaj-aditya) [reply .vscode/launch.json:21]: done

**Comment by bharadwaj-aditya** [comment]
Made the changes

File: main.go (hunk 5 above)
Comment (agasheesh) [reply main.go]: Sounds reasonable, I agree that adding more robust handling as part of subcommands is the way to go forward here.

File: main.go (hunk 7 above)
Comment (agasheesh) [reply main.go]: sounds good.

File: main.go (hunk 6 above)
Comment (agasheesh) [reply main.go:155]: making this as resolved

**Review by agasheesh — APPROVED** [review]
